plsync/mlabconfig_test.py

```diff
@@ -234,6 +295,19 @@ def test_export_mlab_zone_header(self):
 
         self.assertEqual(output.getvalue(), 'before; middle; after')
 
+    @mock.patch('__builtin__.open')
+    def test_export_mlab_server_network_config(self, mock_open):
+        stdout = StringIO.StringIO()
+        name_tmpl = '{{hostname}}-foo.ipxe'
+        input_tmpl = StringIO.StringIO('ip={{ip}} ; echo ${ip}')
+        file_output = StringIO.StringIO()
+        mock_open.return_value = OpenStringIO(file_output)
+
+        mlabconfig.export_mlab_server_network_config(
+            stdout, self.sites, name_tmpl, input_tmpl, 'mlab1.abc01')
+
+        self.assertEqual(file_output.getvalue(), 'ip=192.168.1.9 ; echo ${ip}')
```

Comment:
This the power of `mock`.

This test depends on being able to control what the builtin call to `open` does within the function `mlabconfig.export_mlab_server_network_config`.

The decorator `@mock.patch` replaces the builtin `open` function with a mock object. When the code under test calls `open` it will actually be acting on this mocked object. On line 304, we set the `return_value` for that object to a value we do control.

The `OpenStringIO` function returns a generator, and generators happen to work as a contextmanager in `with` statements. The argument to `OpenStringIO` is the `file_output` that will record all action on the code under test "opened". 

On line 424 from mlabconfig.py, the value `f` is the `file_output` object. The `f.write` writes to the result of the template evaluation and since the unit test has a reference to `file_output`, we can then inspect it's content.

My perspective is in general: real objects are preferable to fakes, fakes are preferable to mocks, and mocks exist for cases when life is more difficult without them. Though because mocks are so powerful they can be overused or used poorly.

Does this help?